File: Binary Tree/二叉树的修改与构造/105.从前序与中序遍历序列构造二叉树.cpp
#include <vector>

using namespace std;
/*
 * @lc app=leetcode.cn id=105 lang=cpp
 *
 * [105] 从前序与中序遍历序列构造二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution
{
public:
    // 切割数组
    TreeNode *traversal(vector<int> &preorder, vector<int> &inorder)
    {
        if (preorder.size() == 0)
        {
            return nullptr;
        }

        // 前序的第一个节点为根节点
        int middle = preorder.front();
        TreeNode *root = new TreeNode(middle);

        if (preorder.size() == 1)
        {
            return root;
        }
        // 前序遍历舍弃掉中间的节点，也就是第一个节点

        int index;
        int isLeft = true;
        vector<int> inOrderLeft;
        vector<int> inOrderRight;

        // 分割中序遍历数组
        for (index = 0; index < inorder.size(); index++)
        {
            if (inorder[index] == middle)
            {
                isLeft = false;
                continue;
            }

            if (isLeft)
            {
                inOrderLeft.push_back(inorder[index]);
            }
            else
            {
                inOrderRight.push_back(inorder[index]);
            }
        }
        // 使用中序遍历的左子树的节点数来切割前序遍历
        int leftSize = inOrderLeft.size();
        vector<int> preOrderLeft;
        vector<int> preOrderRight;

        // 分割前序遍历数组，前序的第一个位数为中间节点，所以要去掉
        // index = 1，从第2个元素开始遍历
        for (int i = 1; i < preorder.size(); i++)
        {
            if (i - 1 < leftSize)
            {
                preOrderLeft.push_back(preorder[i]);
            }
            else
            {
                preOrderRight.push_back(preorder[i]);
            }
        }

        root->left = traversal(preOrderLeft, inOrderLeft);
        root->right = traversal(preOrderRight, inOrderRight);

        return root;
    }

    TreeNode *buildTree(vector<int> &preorder, vector<int> &inorder)
    {
        if (preorder.size() == 0 || inorder.size() == 0)
        {
            return nullptr;
        }

        return traversal(preorder, inorder);
    }
};
// @lc code=end

File: Binary Tree/二叉树的修改与构造/105.从前序与中序遍历序列构造二叉树.ts
/*
 * @lc app=leetcode.cn id=105 lang=typescript
 *
 * [105] 从前序与中序遍历序列构造二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function buildTree(preorder: number[], inorder: number[]): TreeNode | null {
  if (preorder.length === 0) {
    return null
  }

  if (preorder.length === 1) {
    return new TreeNode(preorder[0])
  }

  // 由于前序遍历和中序遍历长度肯定是一样的
  // 所以前序数组不为空，中序数组肯定也不为空
  let top = preorder.shift()
  // 从中序遍历中找到根节点位置
  let index = 0

  for (let i = 0; i < inorder.length; i++) {
    if (inorder[i] === top) {
      // 根节点位置，也就是切割点位置
      index = i
    }
  }
  // 中序遍历，top节点在中间
  // top节点切割中序数组
  // 这里由于index为top的位置，所以需要index+1，过滤掉top节点
  const leftInorder = inorder.slice(0, index)
  const rightInorder = inorder.slice(index + 1)
  // 前序数组去掉top后，也要切割
  // 因为切割之后左右数组长度肯定和中序一样
  // 所以直接使用跟中序数组一样的长度进行切割
  const leftPreorder = preorder.slice(0, index)
  const rightPreorder = preorder.slice(index)

  return new TreeNode(
    top,
    buildTree(leftPreorder, leftInorder),
    buildTree(rightPreorder, rightInorder))
};
// @lc code=end

